torch/nn/functional.py

```diff
@@ -11,121 +11,266 @@
+        the output has shape ``ceil(in_shape / stride)`` in each dimension.
+        For the default ``stride = 1``, the output is the same size as the input.
 
-conv2d = _add_docstr(torch.conv2d, r"""
-conv2d(input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1) -> Tensor
+        .. warning::
+            For ``padding='same'``, if the ``weight`` is even-length and
+            ``dilation`` is odd in any dimension, a full :func:`pad` operation
+            may be needed internally. Lowering performance.
 
-Applies a 2D convolution over an input image composed of several input
-planes.
+    Args:
+        input: input tensor of shape :math:`(\text{minibatch} , \text{in\_channels} , iW)`
+        weight: filters of shape :math:`(\text{out\_channels} , \frac{\text{in\_channels}}{\text{groups}} , kW)`
+        bias: optional bias of shape :math:`(\text{out\_channels})`. Default: ``None``
+        stride: the stride of the convolving kernel. Can be a single number or
+        a one-element tuple `(sW,)`. Default: 1
+        padding: implicit paddings on both sides of the input. Can be a mode {'valid', 'same'},
+        single number or a one-element tuple `(padW,)`. Default: 0
+        dilation: the spacing between kernel elements. Can be a single number or
+        a one-element tuple `(dW,)`. Default: 1
+        groups: split input into groups, :math:`\text{in\_channels}` should be divisible by
+        the number of groups. Default: 1
 
-See :class:`~torch.nn.Conv2d` for details and output shape.
+    Examples::
 
-Note:
-    In some circumstances when using the CUDA backend with CuDNN, this operator
-    may select a nondeterministic algorithm to increase performance. If this is
-    undesirable, you can try to make the operation deterministic (potentially at
-    a performance cost) by setting ``torch.backends.cudnn.deterministic =
-    True``.
-    Please see the notes on :doc:`/notes/randomness` for background.
+        >>> filters = torch.randn(33, 16, 3)
+        >>> inputs = torch.randn(20, 16, 50)
+        >>> F.conv1d(inputs, filters)
+    """
+    if not torch.jit.is_scripting():
+        if ((type(input) is not Tensor or type(weight) is not Tensor) and
+                has_torch_function((input, weight))):
+            return handle_torch_function(
+                conv1d, (input, weight), input, weight, bias, stride,
+                padding, dilation, groups)
+
+    if isinstance(padding, str):
+        if padding == 'same':
+            return _VF._conv1d_same(input, weight, bias, stride, dilation, groups)
+        if padding == 'valid':
+            return _VF.conv1d(input, weight, bias, stride, padding=0,
+                              dilation=dilation, groups=groups)
+        else:
+            raise ValueError('Invalid padding mode {!r}'.format(padding))
+    else:
+        # FIXME: JIT doesn't understand this
+        assert not isinstance(padding, str)
```

Comment:
The JIT won't doesn't support `Union` so instead of `Union[str, BroadcastingList1[int]]` I have to use two `_overload`s with different types for `padding`. This doesn't compile though because, while the JIT understands the `if isinstance` branch must contain a string, it doesn't understand that the `else` branch must _not_ contain a string. So, JIT fails because `_VF.conv1d` doesn't support a string `padding`.

This could be fixed either in the JIT, or by updating the `native_functions.yaml` schema to support string types. That is, unless I'm missing something and there's already a way to support JIT here.

@ezyang any thoughts on this?

Reply:
I don't know. @suo any ideas?